File: ddd-briscola-web/src/main/ts/model/ws/competition.ts
namespace Model.Ws {

  export type CompetitionEvent = CompetitionDeclined | CompetitionAccepted | ConfirmedCompetition | CreatedCompetition

  export interface CompetitionDeclined extends DomainEvent {
    player: Path
    competition: Path
    reason?: string
  }

  export interface CompetitionAccepted extends DomainEvent {
    player: Path
    competition: Path
  }

  export interface ConfirmedCompetition extends DomainEvent {
    competition: Path
  }

  export interface CreatedCompetition extends DomainEvent {
    issuer: Path
    competition: Path
  }

  export interface CompetitionState extends State {
    self?: Path
    kind: string
    competition?: Competition
    acceptingPlayers: Path[]
    decliningPlayers: Path[]
    accept?: Path
    decline?: Path
  }

  export type CompetitionStartDeadline = AllPlayers | OnPlayerCount

  export interface AllPlayers {
    kind: string
  }
  
  export interface OnPlayerCount {
    count: number
    kind: string
  }

  export interface Competition {
    players: Path[]
    kind: MatchKind
    deadline: CompetitionStartDeadline
  }

  export type MatchKind = SingleMatch | NumberOfGamesMatchKind | TargetPointsMatchKind

  export interface SingleMatch {
    kind: string
  }
  
  export interface NumberOfGamesMatchKind {
    numberOfMatches: number
    kind: string
  }

  export interface TargetPointsMatchKind {
    winnerPoints: number
    kind: string
  }

  const competitionEvents: Std.JsMap<boolean> = {
    createdCompetition: true,
    confirmedCompetition: true,
    playerAccepted: true,
    playerDeclined: true
  }

  export function isCompetitionEvent(str: string) {
    return competitionEvents[str] === true;
  }

}

File: ddd-briscola/src/main/scala/com/github/gdefacci/briscola/competition/competition.scala
package com.github.gdefacci.briscola
package competition

import player._
import com.github.gdefacci.ddd._
import scalaz.{ -\/, \/, \/- }
import java.time.LocalDateTime

import com.github.gdefacci.briscola.game._

trait CompetitionDecider extends Decider[CompetitionState, CompetitionCommand, CompetitionEvent, CompetitionError] {

  def nextId: CompetitionId
  def playerById(id: PlayerId): Option[Player]

  def apply(s: CompetitionState, cmd: CompetitionCommand): CompetitionError \/ Seq[CompetitionEvent] = {
    (s -> cmd) match {
      case (EmptyCompetition, CreateCompetition(issuer, gmPlayers, kind, deadLine)) =>
        val gamePlayers: GamePlayers = gmPlayers match {
          case Players(players) => Players(players + issuer)
          case gpls => gpls
        }
        GamePlayersValidator.withValidPlayersAndTeams(gamePlayers, playerById(_)) { (players, teams) =>
          val issr = players.find(_.id == issuer).get
          Seq(CreatedCompetition(nextId, issr, Competition(gamePlayers, kind, deadLine)))
        } leftMap(CompetioBriscolaError(_))
        
      case (comp: OpenCompetition, AcceptCompetition(playerId)) =>
        val players = GamePlayers.getPlayers(comp.competition.players)
        if (comp.acceptingPlayers.contains(playerId)) {
          -\/(CompetitionAlreadyAccepted)
        } else if (players.contains(playerId)) {
          \/-(Seq(CompetitionAccepted(playerId)))
        } else {
          -\/(CompetioBriscolaError(InvalidPlayer(playerId)))
        }
      case (comp: OpenCompetition, DeclineCompetition(playerId, reason)) =>
        val players = GamePlayers.getPlayers(comp.competition.players)
        if (comp.decliningPlayers.contains(playerId)) {
          -\/(CompetitionAlreadyDeclined)
        } else if (players.contains(playerId)) {
          \/-(Seq(CompetitionDeclined(playerId, reason)))
        } else {
          -\/(CompetioBriscolaError(InvalidPlayer(playerId)))
        }
      case (EmptyCompetition, _) => -\/(CompetitionNotStarted)
      case (comp: OpenCompetition, cmd: CreateCompetition) => -\/(CompetitionAlreadyStarted)
      case (c: DroppedCompetition, _) => -\/(CompetitionDropped)
      case (c: FullfilledCompetition, _) => -\/(CompetitionAlreadyFinished)
    }
  }

}

trait CompetitionEvolver extends Evolver[CompetitionState, CompetitionEvent] {

  def isFullfilled(comp: Competition, players: Set[PlayerId]) = {
    val compPlayers = GamePlayers.getPlayers(comp.players)
    comp.deadline match {
      case CompetitionStartDeadline.AllPlayers => compPlayers == players
      case CompetitionStartDeadline.OnPlayerCount(n) => compPlayers.size == n
    }
  }

  def isDropped(comp: Competition, decliningPlayers: Set[PlayerId]) = {
    comp.deadline match {
      case CompetitionStartDeadline.AllPlayers => decliningPlayers.nonEmpty
      case CompetitionStartDeadline.OnPlayerCount(n) => (GamePlayers.getPlayers(comp.players).size - decliningPlayers.size) < n
    }
  }

  def apply(s: CompetitionState, event: CompetitionEvent): CompetitionState = {
    (s -> event) match {
      case (EmptyCompetition, CreatedCompetition(id, issuer, comp)) =>
        OpenCompetition(id, comp, Set(issuer.id), Set.empty)

      case (OpenCompetition(id, competition, acceptingPlayers, decliningPlayers), CompetitionAccepted(pid)) =>
        if (isFullfilled(competition, acceptingPlayers + pid))
          FullfilledCompetition(id, competition, acceptingPlayers + pid, decliningPlayers - pid)
        else
          OpenCompetition(id, competition, acceptingPlayers + pid, decliningPlayers - pid)

      case (OpenCompetition(id, competition, acceptingPlayers, decliningPlayers), CompetitionDeclined(pid, rsn)) =>
        if (isDropped(competition, decliningPlayers + pid))
          DroppedCompetition(id, competition, acceptingPlayers - pid, decliningPlayers + pid)
        else
          OpenCompetition(id, competition, acceptingPlayers - pid, decliningPlayers + pid)

      case _ => {
        throw new RuntimeException(s"forbidden condition state:${s} event:${event}")

      }
    }
  }
}